### PR TITLE
Merge `bitvec.rs` and `indexed_set.rs`

### PR DESCRIPTION
**Pull request by nnethercote**
Because it's not good to have two separate implementations. Also, I will combine the best parts of each to improve NLL memory usage on some benchmarks significantly.

### CODE REVIEW

**Comment by nnethercote** [comment]
> Also, I will combine the best parts of each to improve NLL memory usage on some benchmarks significantly.

See https://github.com/rust-lang/rust/issues/52028#issuecomment-421944193 for some measurements.

File: src/librustc_data_structures/bit_set.rs [hunk 1]
```diff
@@ -0,0 +1,1029 @@
...
+use std::marker::PhantomData;
+use std::mem;
+use std::slice;
+
+pub type Word = u64;
+pub const WORD_BYTES: usize = mem::size_of::<Word>();
+pub const WORD_BITS: usize = WORD_BYTES * 8;
+
+/// A fixed-size bitset type with a dense representation.
+///
```
Comment (pnkfelix) [inline src/librustc_data_structures/bit_set.rs]: The documentation for `struct IdxSet` attempted to explain the relationship between `T` and the `IdxSet<T>`. The doc-comment here does not do that. Would you mind skimming over the doc-comments in the now deleted `indexed_set.rs` and porting over any, like that for `struct IdxSet`, that add useful info for a newcomer trying to understand the type?

File: src/librustc_data_structures/bit_set.rs [hunk 2]
```diff
@@ -160,7 +160,7 @@ impl<T: Idx> BitSet<T> {
     /// Set `self = self & other` and return true if `self` changed.
     /// (i.e., if any bits were removed).
     pub fn intersect(&mut self, other: &BitSet<T>) -> bool {
-        bitwise(self.words_mut(), other.words(), &Intersect)
+        bitwise(self.words_mut(), other.words(), |a, b| { a & b })
```
Comment (pnkfelix) [inline src/librustc_data_structures/bit_set.rs]: hmm I guess I can hope this always gets inlined even on debug builds (which I believe was a property of the former architecture here...)

**Comment by pnkfelix** [comment]
r=me; I'd like some of the deleted documentation to be preserved, but I'm not going to block landing the PR on completion of that task. I'll let @nnethercote decide whether to port over the docs or not as I asked in a comment in my review.

File: src/librustc_data_structures/bit_set.rs (hunk 1 above)
Comment (nnethercote) [reply src/librustc_data_structures/bit_set.rs]: This is the old comment:
```
/// Represents a set of some element type E, where each E is identified by some
/// unique index type `T`.
///
/// In other words, `T` is the type used to index into the bitvector
/// this type uses to represent the set of object it holds.
```
I always found the part about `E` confusing, which is why I dropped the comment. And given that the new type name is `BitSet` it should be clear that the set elements are bits. But I'm happy to add something about `T` being an index type; I can do that for all the data structures in this file.

**Comment by nnethercote** [comment]
I have added comments about the index types, and fixed the problems with the tests.

r=pnkfelix

**Pull request merged**